test/test-try-commit.js

```diff
@@ -0,0 +1,523 @@
+			.then(res => {
+				t.false(res.passed);
+				t.ok(res.errors);
+				t.is(res.errors.length, 1);
+				const error = res.errors[0];
+				t.match(error.message, /Test finished without running any assertions/);
+				t.is(error.name, 'Error');
+				res.commit();
+			});
+	}).run().then(result => {
+		t.false(result.passed);
+	});
+});
+
+test('try-commit fails when no assertions inside multiple try', t => {
+	return ava(a => {
+		return Promise.all([
+			a.try(b => b.pass()).then(res1 => {
+				res1.commit();
+				t.true(res1.passed);
+			}),
+			a.try(() => {}).then(res2 => {
+				t.false(res2.passed);
+				t.ok(res2.errors);
+				t.is(res2.errors.length, 1);
+				const error = res2.errors[0];
+				t.match(error.message, /Test finished without running any assertions/);
+				t.is(error.name, 'Error');
+				res2.commit();
+			})
+		]);
+	}).run().then(result => {
+		t.false(result.passed);
+	});
+});
+
+test('test fails when try-commit committed to failed state', t => {
+	return ava(a => {
+		return a.try(b => b.fail()).then(res => {
+			t.false(res.passed);
+			res.commit();
+		});
+	}).run().then(result => {
+		t.false(result.passed);
+	});
+});
+
+test('try-commit has proper titles, when going in depth and width', t => {
+	t.plan(6);
+	return ava(a => {
+		t.is(a.title, 'test');
+
+		return Promise.all([
+			a.try(b => {
+				t.is(b.title, 'test (attempt 1)');
+
+				return Promise.all([
+					b.try(c => t.is(c.title, 'test (attempt 1) (attempt 1)')),
+					b.try(c => t.is(c.title, 'test (attempt 1) (attempt 2)'))
+				]);
+			}),
+			a.try(b => t.is(b.title, 'test (attempt 2)')),
+			a.try(b => t.is(b.title, 'test (attempt 3)'))
+		]);
+	}).run();
+});
+
+test('try-commit does not fail when calling commit twice', t => {
+	return ava(a => {
+		return a.try(b => b.pass()).then(res => {
+			res.commit();
+			res.commit();
+		});
+	}).run().then(result => {
+		t.true(result.passed);
+		t.false(result.error);
+	});
+});
+
+test('try-commit does not fail when calling discard twice', t => {
+	return ava(a => {
+		return a.try(b => b.pass()).then(res => {
+			res.discard();
+			res.discard();
+		});
+	}).run().then(result => {
+		t.false(result.passed);
+		t.ok(result.error);
+		t.match(result.error.message, /Test finished without running any assertions/);
+		t.is(result.error.name, 'Error');
+	});
+});
+
+test('try-commit allows planning inside the try', t => {
+	return ava(a => {
+		return a.try(b => {
+			b.plan(3);
+
+			b.pass();
+			b.pass();
+			b.pass();
+		}).then(res => {
+			t.true(res.passed);
+			res.commit();
+		});
+	}).run().then(result => {
+		t.true(result.passed);
+	});
+});
+
+test('try-commit fails when plan is not reached inside the try', t => {
+	return ava(a => {
+		return a.try(b => {
+			b.plan(3);
+
+			b.pass();
+			b.pass();
+		}).then(res => {
+			t.false(res.passed);
+			res.commit();
+		});
+	}).run().then(result => {
+		t.false(result.passed);
+	});
+});
+
+test('try-commit passes with failing test', t => {
```

Comment:
What do you mean by this? A failure inside the attempt should cause the main test to fail if that attempt is committed.

Reply:
The title is not clear at all. What I meant is that the test that is marked as failing (expected to fail) will pass, even if try-commit within the test will fail.